### PR TITLE
Add the Session.virtualfile_from_stringio method to allow StringIO input for certain functions/methods

### LINKED ISSUE #571: Allow Figure.legend to read from StringIO
I need to display a legend of symbols for the magnitude of earthquakes. Currently the legend specification must be a file on disk. The Pythonic implementation should allow **Figure.legend(spec=**_<buffer|file>_**,...)**. My preference is to write the legend specification to io.StringIO buffer and then use the buffer for Figure.legend. Instead I have to generate the legend spec in program and write to a file and then specify **spec='my_legend.txt'**.

Figure.legend has default arg **spec=None**. I have no idea how this default works, that's a separate issue: _improve documentation_.

I am willing to implement, but have no experience with such participation.

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This PR adds support for StringIO inputs, which was initially requested in #571 (4 yeas ago!).

GMT doesn't natively support reading StringIO objects. To pass a StringIO object into GMT, we need to create a GMT_DATASET container and store the contents of the StringIO object in the GMT_DATASET container. 

Ideally, any PyGMT methods/functions should accept StringIO input, but it's technically difficult to implement. The main reason is that, when creating a GMT_DATASET container, we need to specify the number of tables, segments, columns, and rows, which means we must try very hard to parse the contents of the StringIO object. The number of columns actually is the most difficult to determine. For example, the content below should be 2 numeric columns and one trailing text column, but we may incorrectly parse it to have 0 numeric columns:
```
30.0W 20.0S TEXT1 TEXT4
35.0W 20.5S TEXT2
```
and the following content that we want to pass to `Figure.text`'s paragraph mode (#1078), should have zero numeric columns and one trailing text column, but we may incorrectly think it has two numeric columns: 
```
gmt text -R0/3/0/5 -JX3i -h1 -M -N -F+f12,Times-Roman+jLT -pdf figure << EOF
This is an unmarked header record not starting with #
> 0 -0.5 13p 3i j
20240101 010101 Event 1
20240101 010101 Event 2
EOF
```
Thus, to keep things as simple as possible, I think we should only allow StringIO input for a few modules (e.g., makes much sense for `Figure.legend`, but makes little sense to `Figure.plot`). In this case, we just need to store the StringIO contents into the trailing text of the GMT_DATASET container.

This PR implements the virtualfile_from_string method to allow StringIO objects. There are two related PRs:

- Support StringIO object in `Figure.legend` (PR #3438 addressing #571)
- Let `Figure.text` support paragraph mode (PR #XX addressing #1078)

**Reminders**

- [x] Run `make format` and `make check` to make sure the code follows the style guide.
- [x] Add tests for new features or tests that would have caught the bug that you're fixing.
- [x] Add new public functions/methods/classes to `doc/api/index.rst`.
- [x] Write detailed docstrings for all functions/methods.
- [ ] If wrapping a new module, open a 'Wrap new GMT module' issue and submit reasonably-sized PRs.
- [ ] If adding new functionality, add an example to docstrings or tutorials.
- [ ] Use underscores (not hyphens) in names of Python files and directories.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash command is:

- `/format`: automatically format and lint the code

### CODE REVIEW

**Comment by seisman** [comment]
> Ideally, any PyGMT methods/functions should accept StringIO input, but it's technically difficult to implement. The main reason is that, when creating a GMT_DATASET container, we need to specify the number of tables, segments, columns, and rows, which means we must try very hard to parse the contents of the StringIO object. The number of columns actually is the most difficult to determine.

Actually, when calling a GMT module, we usually know the expected number of numeric columns (for example, 0 for `legend`, 2 or more for `plot`), so we could add an extra parameter `n_columns` to the `virtualfile_from_stringio` function after refactoring the `virtualfile_in` method in #2744. But I feel that passing a StringIO object to modules like `plot` makes little sense and has very low priority. Then the question is, how can we add StringIO support to specific functions only (e.g., `Figure.legend`)?

File: pygmt/clib/session.py [hunk 1]
```diff
@@ -1603,6 +1610,83 @@ def virtualfile_from_grid(self, grid):
...
+        header = None
+        string_arrays = []
+        for line in stringio.getvalue().splitlines():
+            if line.startswith("#"):  # Skip comments
+                continue
+            if line.startswith(">"):  # Segment header
+                if header is not None:  # Only one segment is allowed now.
+                    raise GMTInvalidInput("Only one segment is allowed.")
+                header = line
+                continue
```
Comment (weiji14) [inline pygmt/clib/session.py:1664]: Need a test to check that multi-segment inputs fail.

File: pygmt/clib/session.py [hunk 2]
```diff
@@ -1603,6 +1610,83 @@ def virtualfile_from_grid(self, grid):
...
+        with self.open_virtualfile(family, geometry, "GMT_IN", dataset) as vfile:
+            try:
+                yield vfile
+            finally:
+                # Must set the text to None to avoid double freeing the memory
+                seg.text = None
+
     def virtualfile_in(  # noqa: PLR0912
         self,
         check_kind=None,
```
Comment (weiji14) [inline pygmt/clib/session.py:1692]: > Then the question is, how can we add StringIO support to specific functions only (e.g., `Figure.legend`)?

Maybe `legend` should have a special check_kind? E.g. `if check_kind("legend"): valid_kinds += ("stringio", ...)`?

Comment (seisman) [reply pygmt/clib/session.py:1692]: See https://github.com/GenericMappingTools/pygmt/pull/3438 for the POC PR for `Figure.legend`. Since we already checked the data kind before entering the session, we can have `check_kind=False` in `Figure.legend`.

File: pygmt/clib/session.py (hunk 1 above)
Comment (seisman) [reply pygmt/clib/session.py:1667]: It's likely we have to allow multi-segments in StringIO.

Here is a CLI version for typesetting a paragraph of text. It seems the first line must be a line not starting with `#` and is ignored. Not sure why it was designed like this. I'll open another POC PR for `Figure.text` and see how it works.
```
gmt text -R0/3/0/5 -JX3i -h1 -M -N -F+f12,Times-Roman+jLT -pdf figure << EOF
This is an unmarked header record not starting with #
> 0 -0.5 13p 3i j
@%5%Figure 1.@%% This illustration shows nothing useful, but it still needs
a figure caption. Highlighted in @;255/0/0;red@;; you can see the locations
of cities where it is @\_impossible@\_ to get any good Thai food; these are to be avoided.
EOF
```

Comment (seisman) [reply pygmt/clib/session.py]: Correction: I didn't see the `-h1` option in the CLI version, which skips the first line. Without `-h1`, the CLI version still works:
```
gmt text -R0/3/0/5 -JX3i -M -N -F+f12,Times-Roman+jLT -pdf figure << EOF
> 0 -0.5 13p 3i j
@%5%Figure 1.@%% This illustration shows nothing useful, but it still needs
a figure caption. Highlighted in @;255/0/0;red@;; you can see the locations
of cities where it is @\_impossible@\_ to get any good Thai food; these are to be avoided.
EOF
````

Comment (seisman) [reply pygmt/clib/session.py]: Here is a POC example to show how to support paragraph mode via StringIO:
```python
In [1]: import io

In [2]: stringio = io.StringIO(
   ...:     "> 0 -0.5 13p 3i j\n"
   ...:     "@%5%Figure 1.@%% This illustration shows nothing useful, but it still needs\n"
   ...:     "a figure caption. Highlighted in @;255/0/0;red@;; you can see the locations\n"
   ...:     "of cities where it is @_impossible@_ to get any good Thai food; these are to be avoided.\n"
   ...: )

In [3]: import pygmt

In [4]: fig = pygmt.Figure()

In [5]: from pygmt.clib import Session

In [6]: with Session() as lib:
   ...:     with lib.virtualfile_in(data=stringio) as vintbl:
   ...:         lib.call_module(module="text", args=f"{vintbl} -R0/3/0/5 -JX3i -M -N -F+f12,Times-Roman+jLT")
   ...:

In [7]: fig.show()
```

Comment (seisman) [reply pygmt/clib/session.py]: StringIO objects containing multi-segments are supported in the latest version.

File: pygmt/clib/session.py [hunk 3]
```diff
@@ -1603,6 +1610,100 @@ def virtualfile_from_grid(self, grid):
...
+            family,
+            geometry,
+            mode="GMT_CONTAINER_ONLY|GMT_WITH_STRINGS",
+            dim=[n_tables, n_segments, n_rows, n_columns],
+        )
+        dataset = ctp.cast(dataset, ctp.POINTER(_GMT_DATASET))
+        table = dataset.contents.table[0].contents
+        for i, segment in enumerate(segments):
+            seg = table.segment[i].contents
+            if segment["header"] != "":
```
Comment (seisman) [inline pygmt/clib/session.py:1692]:
```suggestion
            if segment["header"]:
```

File: pygmt/clib/session.py [hunk 4]
```diff
@@ -60,6 +61,7 @@
     "GMT_IS_PLP",  # items could be any one of POINT, LINE, or POLY
     "GMT_IS_SURFACE",  # items are 2-D grid
     "GMT_IS_VOLUME",  # items are 3-D grid
+    "GMT_IS_TEXT",  # Text strings which triggers ASCII text reading
```
Comment (weiji14) [inline pygmt/clib/session.py:64]: Is only ASCII supported for now? What about other encodings?

**Review by weiji14 — APPROVED** [review]
Thanks @seisman for adding multi-segment support! I just have one concern/thought about non-ASCII character support, but that could be handled later (if needed).

File: pygmt/clib/session.py (hunk 4 above)
Comment (seisman) [reply pygmt/clib/session.py:64]: GMT only accepts ASCII text strings. Any non-ASCII characters should be written in the form of octal codes.
